### DISCUSSION TITLE
Env vars not working

### DISCUSSION THREAD

**Discussion by 619** [vercel/community#4715] [Help] [answered]
### Page to Investigate

https://brokenplatform.com

### Steps to Reproduce

Create new project using create-react-app. Deploy as existing project to vercel. Add environment variables in vercel dashboard for the corresponding project. Update both frontend and backend code and then deploy. Test, not working. Run 'vercel env pull'. Test, not working.

Fix your shit Vercel

**Comment by amyegan** [comment, marked as answer]
Hi @619. I understand that you're frustrated. Please remember to be kind and refrain from using rude language in future posts. 

I just tried the steps you outlined, but I didn't run into any problem with the environment variable I used for testing. 

1. Create new project using `create-react-app`
2. Deploy to Vercel
3. Add `REACT_APP_CLIENT_TEST` to code
4. Add env var in project settings
5. Run `vercel pull && vercel dev` and confirm it's working on local
7. Deploy again: https://react-4715.vercel.app/

Don't forget that environment variables with CRA need the prefix `REACT_APP_`. This is a common reason for missing variables.

> **The environment variables are embedded during the build time.** Since Create React App produces a static HTML/CSS/JS bundle, it can’t possibly read them at runtime. To read them at runtime, you would need to load HTML into memory on the server and replace placeholders in runtime, as [described here](https://create-react-app.dev/docs/title-and-meta-tags#injecting-data-from-the-server-into-the-page). Alternatively you can rebuild the app on the server anytime you change them.
> 
> > Note: You must create custom environment variables beginning with `REACT_APP_`. Any other variables except `NODE_ENV` will be ignored to avoid accidentally [exposing a private key on the machine that could have the same name](https://github.com/facebook/create-react-app/issues/865#issuecomment-252199527).

https://create-react-app.dev/docs/adding-custom-environment-variables/
